onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -0,0 +1,109 @@
+              "Unsupported transformer level specified", level);
+
+  levels_enabled = TransformerLevel::Default;
+  if (level == 1) {
+    levels_enabled |= TransformerLevel::Level1;
+  } else if (level == 2) {
+    levels_enabled |= TransformerLevel::Level1 | TransformerLevel::Level2;
+  }
+
+  if (all_levels != nullptr) {
+    all_levels->push_back(TransformerLevel::Level2);
+    all_levels->push_back(TransformerLevel::Level1);
+  }
+}
+
+std::vector<std::unique_ptr<RewriteRule>> GenerateRewriteRules(const TransformerLevel& level, 
+                                                               const std::vector<std::string>* custom_list) {
+  std::vector<std::unique_ptr<RewriteRule>> rules;
+  switch (level) {
+    case TransformerLevel::Level1:
+      rules.push_back(std::make_unique<EliminateIdentity>());
+      rules.push_back(std::make_unique<EliminateSlice>());
+      break;
+
+    case TransformerLevel::Level2:
+      break;
+    default:
+      ORT_ENFORCE(false, "Unsupported level" + level);
+  }
+
+  if (custom_list != nullptr && !custom_list->empty()) {
+    std::vector<std::unique_ptr<RewriteRule>> filtered_list;
+    for (const auto& rule_name : *custom_list) {
+      std::for_each(rules.begin(), rules.end(), [&](std::unique_ptr<RewriteRule>& item) {
+        if((item != nullptr) && (item->Name() == rule_name)) {
+          filtered_list.push_back(std::move(item));
+        }
+      });
+    }
+    return filtered_list;
+  }
+
+  return rules;
+}
+
+std::vector<TransformerProviderSet> GenerateTransformers(const TransformerLevel& level, 
+                                                         const std::vector<std::string>* custom_list) {
+  std::vector<TransformerProviderSet> transformers;
+  switch (level) {
+    case TransformerLevel::Level1: {
+      std::vector<std::string> l1_execution_providers = {"", onnxruntime::kCpuExecutionProvider};
+      transformers.emplace_back(std::make_unique<UnsqueezeElimination>(), l1_execution_providers);
+    } break;
+
+    case TransformerLevel::Level2: {
+      std::vector<std::string> l2_execution_providers = {onnxruntime::kCpuExecutionProvider};
+      transformers.emplace_back(std::make_unique<ConvAddFusion>(), l2_execution_providers);
+      transformers.emplace_back(std::make_unique<ConvMulFusion>(), l2_execution_providers);
+    } break;
+
+    default:
+      ORT_ENFORCE(false, "Unsupported level" + level);
+      break;
+  }
+
+  if (custom_list != nullptr && !custom_list->empty()) {
+    // pick custom transformers enabled for this session
+    std::vector<TransformerProviderSet> filtered_list;
+    for (const auto& t_name : *custom_list) {
+      std::for_each(transformers.begin(), transformers.end(), 
+          [&](TransformerProviderSet& item) {
+             if((item.first != nullptr) && (item.first->Name() == t_name)){
+               filtered_list.push_back(std::move(item));
+             }
+          });
+    }
+    return filtered_list;
```

Comment:
Rather than return either a "filtered_list" or a "trasformers", it will be easier to read your code if you have a single list that you manipulate. If you have custom_list, you remove some of the transformers.
Or it might be even better to have an "addTransformer" that does at the same time the check if it appears in the custom list. So no need for special loop like this.

Reply:
since custom list size in most cases is smaller or equal to generated rules so I decided to iterate over custom and select rules as opposed to iterate over rules and remove them... in this case since I am not copying the objects it should not matter and what I have is not going to be inefficient 
+ we don't have as many transformers right now to make a perf difference in this loop
+ we only do this once during initialization
+ in this case we are enabling a few transformers so it makes more sense to pick them where as what you are suggesting makes more sense to me when we have a list of names to disable

Reply:
Again I do agree that this is not a critical piece of code, but I don't agree with your explanation, since you will be iterating over all transformers for each transformer in the filtered_list.

Reply:
Same as above.